src/tools/immutable_object_cache/SimplePolicy.cc

```diff
@@ -0,0 +1,215 @@
+
+  m_cache_size = 0;
+
+}
+
+SimplePolicy::~SimplePolicy() {
+  ldout(cct, 20) << dendl;
+
+  for (auto it : m_cache_map) {
+    Entry* entry = (it.second);
+    delete entry;
+  }
+}
+
+cache_status_t SimplePolicy::alloc_entry(std::string file_name) {
+  ldout(cct, 20) << "alloc entry for: " << file_name << dendl;
+
+  RWLock::WLocker wlocker(m_cache_map_lock);
+
+  // cache hit when promoting
+  if (m_cache_map.find(file_name) != m_cache_map.end()) {
+    ldout(cct, 20) << "object is under promoting: " << file_name << dendl;
+    return OBJ_CACHE_SKIP;
+  }
+
+  if ((m_cache_size < m_max_cache_size) &&
+      (inflight_ops < m_max_inflight_ops)) {
+    Entry* entry = new Entry();
+    ceph_assert(entry != nullptr);
+    m_cache_map[file_name] = entry;
+    wlocker.unlock();
+    update_status(file_name, OBJ_CACHE_SKIP);
+    return OBJ_CACHE_NONE;  // start promotion request
+  }
+
+  // if there's no free entry, return skip to read from rados
+  return OBJ_CACHE_SKIP;
+}
+
+cache_status_t SimplePolicy::lookup_object(std::string file_name) {
+  ldout(cct, 20) << "lookup: " << file_name << dendl;
+
+  RWLock::RLocker rlocker(m_cache_map_lock);
+
+  auto entry_it = m_cache_map.find(file_name);
+  // simply promote on first lookup
+  if (entry_it == m_cache_map.end()) {
+      rlocker.unlock();
+      return alloc_entry(file_name);
+  }
+
+  Entry* entry = entry_it->second;
+
+  if (entry->status == OBJ_CACHE_PROMOTED) {
+    // bump pos in lru on hit
+    m_promoted_lru.lru_touch(entry);
+  }
+
+  return entry->status;
+}
+
+void SimplePolicy::update_status(std::string file_name,
+                                 cache_status_t new_status, uint64_t size) {
+  ldout(cct, 20) << "update status for: " << file_name
+                 << " new status = " << new_status << dendl;
+
+  RWLock::WLocker locker(m_cache_map_lock);
+
+  auto entry_it = m_cache_map.find(file_name);
+  if (entry_it == m_cache_map.end()) {
+    return;
+  }
+
+  ceph_assert(entry_it != m_cache_map.end());
+  Entry* entry = entry_it->second;
+
+  // to promote
+  if (entry->status == OBJ_CACHE_NONE && new_status== OBJ_CACHE_SKIP) {
+    entry->status = new_status;
+    entry->file_name = file_name;
+    inflight_ops++;
+    return;
+  }
+
+  // promoting done
+  if (entry->status == OBJ_CACHE_SKIP && new_status== OBJ_CACHE_PROMOTED) {
+    m_promoted_lru.lru_insert_top(entry);
+    entry->status = new_status;
+    entry->size = size;
+    m_cache_size += entry->size;
+    inflight_ops--;
+    return;
+  }
+
+  // promoting failed
+  if (entry->status == OBJ_CACHE_SKIP && new_status== OBJ_CACHE_NONE) {
+    // mark this entry as free
+    entry->file_name = "";
+    entry->status = new_status;
+
+    m_cache_map.erase(entry_it);
+    inflight_ops--;
+    delete entry;
+    return;
+  }
+
+  // to evict
+  if (entry->status == OBJ_CACHE_PROMOTED && new_status== OBJ_CACHE_NONE) {
+    // mark this entry as free
+    uint64_t size = entry->size;
```

Comment:
Name collision with third argument of this method (see line 87). In fact I didn't know this could actually compile...